// use-ssr/dist/useSSR.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
const react_1 = require("react");
var Device;
(function (Device) {
    Device["BROWSER"] = "BROWSER";
    Device["SERVER"] = "SERVER";
    Device["NATIVE"] = "NATIVE";
})(Device = exports.Device || (exports.Device = {}));
const { BROWSER, SERVER, NATIVE } = Device;
const canUseDOM = !!(typeof window !== 'undefined' &&
    window.document &&
    window.document.createElement);
const canUseNative = typeof navigator != 'undefined' && navigator.product == 'ReactNative';
const location = canUseNative ? NATIVE : canUseDOM ? BROWSER : SERVER;
function useSSR() {
    const [whereAmI, setWhereAmI] = react_1.useState(location);
    const mounted = react_1.useRef(false);
    react_1.useEffect(() => {
        if (mounted.current)
            return;
        mounted.current = true;
        setWhereAmI(location);
    });
    const useSSRObject = react_1.useMemo(() => ({
        isBrowser: whereAmI === BROWSER,
        isServer: whereAmI === SERVER,
        isNative: whereAmI === NATIVE,
        canUseWorkers: typeof Worker !== 'undefined',
        canUseEventListeners: whereAmI === BROWSER && !!window.addEventListener,
        canUseViewport: whereAmI === BROWSER && !!window.screen
    }), [whereAmI]);
    return react_1.useMemo(() => Object.assign(Object.values(useSSRObject), useSSRObject), [whereAmI]);
}
exports.default = useSSR;
